fairworkflows/fairstep.py

```diff
@@ -66,24 +68,52 @@ def load_from_nanopub(self, uri):
             if len(concepts_introduced) == 0:
                 raise ValueError('This nanopub does not introduce any concepts. Please provide URI to the step itself (not just the nanopub).')
             elif len(concepts_introduced) > 0:
-                step_uri = concepts_introduced[0]
+                step_uri = str(concepts_introduced[0])
 
             print('Assuming step URI is', step_uri)
 
         else:
             step_uri = uri
 
         self._uri = step_uri
-        self.self_ref = rdflib.URIRef(self._uri)
 
         # Check that the nanopub's assertion actually contains triples refering to the given step's uri
-        if (rdflib.URIRef(self.self_ref), None, None) not in np.assertion:
+        if (rdflib.URIRef(self._uri), None, None) not in np.assertion:
             raise ValueError(f'No triples pertaining to the specified step (uri={step_uri}) were found in the assertion graph of the corresponding nanopublication (uri={np_uri})')
 
         # Else extract all triples in the assertion into the rdf graph for this step
         self._rdf = rdflib.Graph()
         self._rdf += np.assertion
 
+        # Record that this RDF originates from a published source
+        self._is_published = True
+
+
+    def publish_as_nanopub(self):
+        """
+        Publishes the rdf for this FairStep as a nanopublication.
+        Returns True if published successfully.
+        """
+
+        # If this step has been modified from a previously publised step, include this in the derived_from PROV (if applicable)
+        derived_from = None
+        if self._is_published is True:
+            if self.is_modified is True:
+                derived_from = self._uri
+            else:
+                warnings.warn(f'Cannot publish() FairStep. This step is already published (at {self._uri}) and has not been modified.')
+                return False
+
+        # Publish the rdf of this step as a nanopub
+        np_uri = Nanopub.publish(self._rdf, introduces_concept=self.self_ref, derived_from=derived_from)
+
+        # Set the new (published) URI of this fair step, which should be the nanopub URI plus a fragment given by the name of self.self_ref
+        self._uri = np_uri + '#' + str(self.self_ref)
```

Comment:
Yes this is kind of awkward. The problem is that the nanopublication infrastructure only cares about the nanopub and the URI of that nanopub. I originally would use the nanopublication's URI as the step URI, since its assertion graph contained triples describing it, but this was rightfully objected to by the semantic modellers. The step (or workflow) is distinct from the nanopub that describes it, and it must therefore have its own URI. This is generally the nanopub URI with a fragment, such as ```#step``` appended.

The assertion graph could also in principle contain descriptions for several steps, or even the RDF of a whole workflow. So we have to be doubly careful that distinct objects are given their own URI.

Nanopubs have an ```npx:introduces``` predicate for stating that the nanopub introduces a new thing or concept. For example, it might introduce a description of a step (rather than just referring to an existing one). But a nanopub could introduce more than one, perhaps.

What is happening above, is that ```Nanopub.publish()``` is returning where it has published the nanopublication we asked it to. It does not know what URIs or fragments we used to describe one (or more) concepts within the nanopub. So I'm constructing the step's URI based on knowing that the fragment will be the name of the blank node used as the step's root (by default, that is simply 'step'). It's pretty hacky, and I agree it is unclear. I'm just not sure how to make it less so. I want to keep the ```Nanopub``` module as its own standalone thing, and not force it to know about what weird stuff we're stuffing into its assertion graph. Perhaps as you suggest the best way is to have our own little utility function with lots of documentation...?

Reply:
Great, this made it a lot clearer! And therefore indeed I suggest to add this documentation to the code (even though it might become verbose, verbose is better than unclear).

Reply:
Right, I've made Nanopub.publish() now return a dict, instead of just a published URI. This ```publication_info``` dict contains the URI the nanopub was published to, as before, but also includes the URI of the concept that was introduced by the nanopub, if applicable. So the hacky URI construction has at least been moved to the module that determines it, which should make maintenance easier. I've also added a long-ish comment explaining what's happening.